lxd/device/device_utils_unix.go

```diff
@@ -125,109 +157,105 @@ func UnixCreateDevice(s *state.State, idmapSet *idmap.IdmapSet, devicesPath stri
 		if err != nil {
 			return nil, fmt.Errorf("Failed to get device attributes for %s: %s", m["path"], err)
 		}
 	} else if m["major"] == "" || m["minor"] == "" {
 		return nil, fmt.Errorf("Both major and minor must be supplied for device: %s", m["path"])
 	} else {
-		major, err = strconv.Atoi(m["major"])
+		d.Major, err = strconv.Atoi(m["major"])
 		if err != nil {
 			return nil, fmt.Errorf("Bad major %s in device %s", m["major"], m["path"])
 		}
 
-		minor, err = strconv.Atoi(m["minor"])
+		d.Minor, err = strconv.Atoi(m["minor"])
 		if err != nil {
 			return nil, fmt.Errorf("Bad minor %s in device %s", m["minor"], m["path"])
 		}
 	}
 
-	// Get the device mode
-	mode := os.FileMode(0660)
+	// Get the device mode (defaults to 0660 if not supplied).
+	d.Mode = os.FileMode(0660)
 	if m["mode"] != "" {
-		tmp, err := unixGetDeviceModeOct(m["mode"])
+		tmp, err := unixDeviceModeOct(m["mode"])
 		if err != nil {
 			return nil, fmt.Errorf("Bad mode %s in device %s", m["mode"], m["path"])
 		}
-		mode = os.FileMode(tmp)
+		d.Mode = os.FileMode(tmp)
 	} else if !defaultMode {
-		mode, err = shared.GetPathMode(srcPath)
+		d.Mode, err = shared.GetPathMode(srcPath)
 		if err != nil {
 			errno, isErrno := shared.GetErrno(err)
 			if !isErrno || errno != unix.ENOENT {
 				return nil, fmt.Errorf("Failed to retrieve mode of device %s: %s", m["path"], err)
 			}
-			mode = os.FileMode(0660)
+			d.Mode = os.FileMode(0660)
 		}
 	}
 
 	if m["type"] == "unix-block" {
-		mode |= unix.S_IFBLK
+		d.Mode |= unix.S_IFBLK
+		d.Type = "b"
 	} else {
-		mode |= unix.S_IFCHR
+		d.Mode |= unix.S_IFCHR
+		d.Type = "c"
 	}
 
-	// Get the device owner
-	uid := 0
-	gid := 0
-
+	// Get the device owner.
 	if m["uid"] != "" {
-		uid, err = strconv.Atoi(m["uid"])
+		d.UID, err = strconv.Atoi(m["uid"])
 		if err != nil {
 			return nil, fmt.Errorf("Invalid uid %s in device %s", m["uid"], m["path"])
 		}
 	}
 
 	if m["gid"] != "" {
-		gid, err = strconv.Atoi(m["gid"])
+		d.GID, err = strconv.Atoi(m["gid"])
 		if err != nil {
 			return nil, fmt.Errorf("Invalid gid %s in device %s", m["gid"], m["path"])
 		}
 	}
 
-	// Create the devices directory if missing
+	// Create the devices directory if missing.
 	if !shared.PathExists(devicesPath) {
 		os.Mkdir(devicesPath, 0711)
 		if err != nil {
 			return nil, fmt.Errorf("Failed to create devices path: %s", err)
 		}
 	}
 
-	destPath := m["path"]
-	if destPath == "" {
-		destPath = m["source"]
-	}
+	destPath := unixDeviceDestPath(m)
 	relativeDestPath := strings.TrimPrefix(destPath, "/")
-	devName := fmt.Sprintf("%s.%s", strings.Replace(prefix, "/", "-", -1), strings.Replace(relativeDestPath, "/", "-", -1))
+	devName := unixDeviceEncode(unixDeviceJoinPath(prefix, relativeDestPath))
 	devPath := filepath.Join(devicesPath, devName)
 
-	// Create the new entry
+	// Create the new entry.
 	if !s.OS.RunningInUserNS {
-		encodedDeviceNumber := (minor & 0xff) | (major << 8) | ((minor & ^0xff) << 12)
```

Comment:
@stgraber one thing I did want to check with you on this PR, is this part of the code, where the device number is generated to be passed to `unix.Mknod()`. Currently this uses a custom (somewhat complex and opaque) bit shifting operation to generate the encodedDeviceNumber variable. After researching this a bit, I discovered that there is a built in function in the `unix` package that does the same operation (and is documented as to what the bit shifting is). That function is `unix.Mkdev` and in my testing it does result in the same number coming out as the old version had.

However I wasn't certain that I've got the data type conversion correct, the function takes 2 uint32s and returns a uint64 (even though AFAIK the device number is a 32 bit number). I was thinking that having the d.Major and d.Minor variables as native signed `int` types might cause issues on a 32bit platform if the device number was in the higher end of the 32 bit range. Perhaps I should use ParseUint to ensure the input is always unsigned.

Reply:
Here is an example of a similar implementation https://github.com/moby/moby/blob/master/pkg/system/mknod.go

Reply:
Our switch to unix instead of syscall is pretty recent, so my guess is that syscall just didn't have that function before.
We've seen a number of other cases where we've got our own implementation of logic that can now be found in unix and should just switch over as needed.

I would indeed store Major and Minor as uint32 directly, we shouldn't have to care about negative values and can use ParseUint to get the right value.

Reply:
@stgraber thanks, I've updated that now and removed the `Major()` and `Minor()` functions from the `shared` package too. This should make LXD use uint32 for major/minor values consistently. I've added the changes as a series of separate commits to make it easier to review.